packages/@uppy/tus/src/index.js

```diff
@@ -229,9 +230,9 @@ module.exports = class Tus extends BasePlugin {
             done()
             return () => {}
           })
-          return p
+          return Promise.all([p, userProvidedPromise])
```

Comment:
```suggestion
          // RateLimitedQueue should be in charge of pausing requests, we don't want to duplicate that.
          // Therefor we create an empty promise and resolve that when the queue with the current request is ran.
          // This means we can hold onBeforeRequest here with a Promise.all,
          // together with the user provided onBeforeRequest promise. 
          return Promise.all([p, userProvidedPromise])
```

Reply:
```suggestion
          // If the request has been requeued because it was rate limited by the
          // remote server, we want to wait for `RateLimitedQueue` to dispatch
          // the re-try request.
          // Therefore we create a promise that the queue will resolve when
          // enough time has elapsed to expect not to be rate-limited again.
          // This means we can hold onBeforeRequest here with a `Promise.all`,
          // together with the returned value of the user provided
          // `onBeforeRequest` option callback. 
          return Promise.all([p, userProvidedPromise])
```

Reply:
Nice. I made another tweak in case we want to merge #3720 

```suggestion
          // If the request has been queued because the response status code is in `retryStatusCodes`,
          // we want to wait for `RateLimitedQueue` to dispatch the retry request.
          // Therefore we create a promise which the queue will resolve when enough time has elapsed.
          // This means we can hold the Tus retry here with a `Promise.all`,
          // together with the returned value of the user provided
          // `onBeforeRequest` (in case it returns a promise). 
          return Promise.all([p, userProvidedPromise])
```

Reply:
@Murderlon Should I go with my own suggestion given that https://github.com/transloadit/uppy/pull/3720 has taken a different direction?

Reply:
Sounds good